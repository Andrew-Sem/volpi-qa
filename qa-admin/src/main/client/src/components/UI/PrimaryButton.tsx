import {ButtonHTMLAttributes, FC} from 'react';

export const PrimaryButton: FC<ButtonHTMLAttributes<HTMLButtonElement>> = ({title, ...props}) => {
    return (
        <button className={"bg-indigo-700 px-4 py-2 rounded-lg hover:bg-indigo-600 w-full"} {...props}>
            {title}
        </button>
    );
}
