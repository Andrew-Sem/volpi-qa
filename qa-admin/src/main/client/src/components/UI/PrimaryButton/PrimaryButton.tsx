import {ButtonHTMLAttributes, FC} from 'react';
import {getBtnClassesByVariant} from "../../../utils/getbtnClassesByVariant";

interface PrimaryButtonProps extends ButtonHTMLAttributes<HTMLButtonElement>{
    variant?: "danger" | "safe" | "default"
}

export const PrimaryButton: FC<PrimaryButtonProps> = ({children, className, variant = "default", ...props}) => {
    className += getBtnClassesByVariant(variant, "primary")
    if (!className?.includes("w")) className += " w-full"
    return (
        <button
            className={`${className ? className : ""} px-4 py-2 rounded-lg border text-btn-foreground text-sm`} {...props}>
            {children}
        </button>
    );
}
