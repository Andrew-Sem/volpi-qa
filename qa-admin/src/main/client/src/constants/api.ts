export const API_URL = "localhost:8080/api/v1/admin"
