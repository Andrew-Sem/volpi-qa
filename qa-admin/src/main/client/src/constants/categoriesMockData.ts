import {ICategory} from "../types/ICategory";

export const categories: ICategory[] = [
    {
        id: 1124, name: "first category", questions: [
            {
                id: 112930,
                text: "first question",
                answer: "first answer",
                category: {id: 1124, name: "first category"}
            }
        ]
    },
    {
        id: 22118, name: "second category", questions: [
            {
                id: 123252,
                text: "second question",
                answer: "second answer",
                category: {id: 22118, name: "second category"}
            }
        ]
    }
]
