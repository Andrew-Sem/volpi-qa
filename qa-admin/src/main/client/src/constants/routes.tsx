import {lazy, ReactElement} from "react";
import {Login} from "../pages/Login";
const Categories = lazy(() => import("../pages/Categories"))
const CategoryName = lazy(() => import("../pages/CategoryName"))

interface IRoute {
    path: string
    element: ReactElement
}

export const notAuthorizedRoutes: IRoute[] = [
    {path: "/login", element: <Login/>}
]

export const authorizedRoutes: IRoute[] = [
    {path: "/", element: <Categories/>},
    {path: "/categories/:name", element: <CategoryName/>}
]
