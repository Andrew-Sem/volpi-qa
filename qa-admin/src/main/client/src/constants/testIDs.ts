export const CategoriesListTestID = "CategoriesList-testID"

export const InputTestID = "Input-testID"

export const PrimaryButtonTestID = "PrimaryButton-testID"

export const CategoryRowTestID = "CategoryRow-testID"
