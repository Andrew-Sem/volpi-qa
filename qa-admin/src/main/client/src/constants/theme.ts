import {type Theme} from "../types/Theme";
import {ComputerDesktopIcon, MoonIcon, SunIcon} from "@heroicons/react/24/outline";

export type Themes = {
    light: Theme,
    dark: Theme,
    system: Theme
}

export const themes:Themes = {
    light: {id: "1", title: 'light', Icon: SunIcon},
    dark: {id: "2", title: 'dark', Icon: MoonIcon},
    system: {id: "3", title: 'system', Icon: ComputerDesktopIcon}
}