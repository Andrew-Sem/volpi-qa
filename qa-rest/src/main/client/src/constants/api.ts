export const API_URL = import.meta.env.VITE_REST_API_URL || "http://localhost:8081/api/v1"
