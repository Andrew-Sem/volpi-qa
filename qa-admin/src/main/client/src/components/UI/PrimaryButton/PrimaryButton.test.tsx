import {afterEach, beforeEach, describe, expect, it, vi} from "vitest";
import {render, RenderResult, screen} from "@testing-library/react";
import {PrimaryButton} from "./PrimaryButton";
import userEvent from "@testing-library/user-event";
import {PrimaryButtonTestID} from "../../../constants/testIDs";

describe("PrimaryButton", () => {
    let buttonElement: HTMLButtonElement
    const buttonTitle = "42"
    let renderOption: RenderResult
    const buttonOnClick = vi.fn()
    const buttonOnHover = vi.fn()
    const buttonOnBlur = vi.fn()

    describe("default behaviour", () => {
        beforeEach(() => {
            renderOption = render(
                <PrimaryButton
                    data-testid={PrimaryButtonTestID}
                >
                    {buttonTitle}
                </PrimaryButton>)
            buttonElement = screen.getByTestId(PrimaryButtonTestID)
        })

        it("renders correctly with default and additional classes", () => {
            expect(buttonElement).toHaveClass("bg-primary px-4 py-2 rounded-lg hover:bg-primaryHov w-full max-w-[250px]")
            renderOption.rerender(
                <PrimaryButton
                    data-testid={PrimaryButtonTestID}
                    className={"bg-secondary text-contrast"}
                >
                    {buttonTitle}
                </PrimaryButton>
            )
            expect(buttonElement).toHaveClass("bg-secondary text-contrast")
        })

        it("renders correctly default title", () => {
            expect(buttonElement).toHaveTextContent(buttonTitle)
            expect(buttonElement).toMatchSnapshot()
        })
    })
    describe("user iteraction", () => {
        beforeEach(() => {
            render(
                <PrimaryButton
                    data-testid={PrimaryButtonTestID}
                    onClick={buttonOnClick}
                    onMouseLeave={buttonOnBlur}
                    onMouseEnter={buttonOnHover}
                >
                    {buttonTitle}
                </PrimaryButton>)
            buttonElement = screen.getByTestId(PrimaryButtonTestID)
        })

        afterEach(() => {
            vi.clearAllMocks()
        })

        it("handle click correctly", async () => {
            await userEvent.click(buttonElement)
            expect(buttonOnClick).toHaveBeenCalledOnce()
        })

        it("hover and blur effects works correctly", async () => {
            await userEvent.hover(buttonElement)
            expect(buttonOnHover).toHaveBeenCalledOnce()
            expect(buttonOnBlur).not.toBeCalled()

            await userEvent.unhover(buttonElement)
            expect(buttonOnBlur).toHaveBeenCalledOnce()
        })
    })
})
