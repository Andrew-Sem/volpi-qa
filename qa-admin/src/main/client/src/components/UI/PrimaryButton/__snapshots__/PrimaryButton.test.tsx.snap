// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PrimaryButton > default behaviour > renders correctly default title 1`] = `
<button
  class=" bg-accent hover:bg-accent/80 w-full px-4 py-2 rounded-lg text-btn-foreground text-sm duration-2000"
  data-testid="PrimaryButton-testID"
>
  42
</button>
`;
